### PR TITLE
Issue 5351 - author DC improvements

### LINKED ISSUE #5351: Improving Author DC
### Description

Would be good to have it in author DC drop-down list

image
image
image

### Other plugins or websites examples

_No response_

### How to implement it

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by myroslavve**
# Description
Added missing fields to "author" in DC, and created "customer" type for fields from issue that are added by WooCommerce.
Added "current" option both for authors and customers:
For "author" it shows the author of the current post, and for "customer" it shows the user that is viewing the page.

Fixes #5351 

# How Has This Been Tested?
Checked that all new fields work both on frontend and editor. Checked that current option works, and also checked it on FSE.

# Test checklist

**_ Front/Back Testing _**

-   [ ] Check that all new fields work both on frontend and editor.
-   [ ] Check that current option works, also check it on FSE.

## Summary by CodeRabbit

- **New Features**
	- Enhanced dynamic content rendering with new link targets for authors and customers.
	- Introduced a new property in the `LinkControl` component to disable the "Open in new tab" toggle.
	- Added a new method for responsive class updates in the `MaxiBlockComponent`.

- **Bug Fixes**
	- Improved handling of user data retrieval and link generation logic.

- **Style**
	- Added a new modifier class for disabled toggle switches to improve accessibility.

- **Documentation**
	- Updated constants for user-related links and attributes.

- **Chores**
	- General code optimizations and refactoring for improved readability and maintainability.

### CODE REVIEW

**Comment by elzadj** [comment]
@myroslavve also pls check this one https://github.com/maxi-blocks/maxi-blocks/issues/5691#issuecomment-2304908093

**Comment by myroslavve** [comment]
> @myroslavve also pls check this one [#5691 (comment)](https://github.com/maxi-blocks/maxi-blocks/issues/5691#issuecomment-2304908093)

Should be fixed here :+1:

File: src/extensions/DC/getDCContent.js [hunk 1]
```diff
@@ -142,6 +142,20 @@ const getDCContent = async (dataRequest, clientId) => {
...
+		if (field === 'billing_name' || field === 'shipping_name') {
+			const firstNameField = field.replace('_name', '_first_name');
+			const firstName = getCustomerDataField(firstNameField);
+			const lastNameField = field.replace('_name', '_last_name');
+			const lastName = getCustomerDataField(lastNameField);
+			return `${firstName} ${lastName}`;
+		}
+
+		return getCustomerDataField(field) || null;
+	}
```
[the remark this answers is not in the document]
Comment (elzadj) [reply src/extensions/DC/getDCContent.js:157]: That makes sense to do, yep.

**Review by Olekrut — CHANGES_REQUESTED** [review]
Simple editor page and FSE were created on master, 
[simple editor.txt](https://github.com/user-attachments/files/16837995/simple.editor.txt)
[FSE code.txt](https://github.com/user-attachments/files/16837998/FSE.code.txt)


1) email - guess DC link should be mailto: current email, not author link
    website - guess, DC link should be to current website, not author link
    image
2) Message on frontend ( FSE )
image
3) Is should be current in simple editor?
image
4) Maybe better to change label to Biographical info or Bio? ( like in user )
image

**Comment by elzadj** [comment]
> website - guess, DC link should be to current website, not author link
Should be to the website from the field, but maybe it's better to add it as an option. 
Same for the email.

Also about the email - we need to at least  add the antispambot function for frontend (it's not a great one, but at least something) https://developer.wordpress.org/reference/functions/antispambot/

**Comment by myroslavve** [comment]
> 1. email - guess DC link should be mailto: current email, not author link
>    website - guess, DC link should be to current website, not author link

Added email and site options to link settings.

> 2. Message on frontend ( FSE )

Fixed.

> 3. Is should be current in simple editor?
>    image

Don't see why not. Seems to be working as I would expect.

> 4. Maybe better to change label to Biographical info or Bio? ( like in user )

Changed.

**Comment by myroslavve** [comment]
> > website - guess, DC link should be to current website, not author link
> > Should be to the website from the field, but maybe it's better to add it as an option.
> > Same for the email.
> 
> Also about the email - we need to at least add the antispambot function for frontend (it's not a great one, but at least something) https://developer.wordpress.org/reference/functions/antispambot/

So nowadays most of the web crawlers seem to work around the wordpress antispambot method, so implemented antispam using javascript to write email on page load, assuming that majority of web crawlers don't execute js, but it still is possible for them to get the email, so added a warning on editor to tell users alternatives to sharing their email on the site:
image

**Comment by elzadj** [comment]
@myroslavve weird how the performance-current fails here, could you please isolate it, and check what's going on? Thanks.

**Comment by elzadj** [comment]
@myroslavve ignore the performance tests for now, I'm going to change that one page that randomly breaks, and change the tests iterations number.

**Review by Olekrut — CHANGES_REQUESTED** [review]
https://jam.dev/c/9737cbf1-344d-4694-b570-c93eb14efa41
Create on master author DC blocks ( some of them with links )
FSE and simple editor have errors and broken blocks.
Just blog pattern with broken blocks as well.
image
image

Everything is ok if create blocks in this branch

**Comment by myroslavve** [comment]
> https://jam.dev/c/9737cbf1-344d-4694-b570-c93eb14efa41 Create on master author DC blocks ( some of them with links ) FSE and simple editor have errors and broken blocks. Just blog pattern with broken blocks as well. image image
> 
> Everything is ok if create blocks in this branch

Fixed validation errors on editor, but I couldn't reproduce the errors from php on frontend, if you still have them after this update, could you share code editor with me please.

**Comment by elzadj** [comment]
@Olekrut

**Review by Olekrut — CHANGES_REQUESTED** [review]
1) FSE - 
[FSE DC author.txt](https://github.com/user-attachments/files/17044177/FSE.DC.author.txt)

image

2) only this branch
image

**Comment by myroslavve** [comment]
> 1. FSE -
>    [FSE DC author.txt](https://github.com/user-attachments/files/17044177/FSE.DC.author.txt)
> 
> image

Fixed.

> 2. only this branch
>    image

This error and warnings are all from Jam, was testing and didn't get any of them once. Maybe try to update Jam. I am 95% sure that changes on this branch shouldn't break anything for Jam.

**Review by Olekrut — APPROVED** [review]

File: core/class-maxi-dynamic-content.php [hunk 2]
```diff
@@ -904,20 +904,14 @@
...
-                    $link = get_author_posts_url($attributes['dc-author']);
+            $dc_link_target = $attributes['dc-link-target'];
+            if (!empty($post) && isset($post->ID)) {
+                if ($dc_link_target === 'author_email') {
+                    $link = $this->xor_obfuscate_email(get_the_author_meta('user_email', $post->ID));
+                } elseif ($dc_link_target === 'author_site') {
+                    $link = get_the_author_meta('user_url', $post->ID);
                 } else {
-                    $link = get_author_posts_url($attributes['dc-id']);
+                    $link = get_author_posts_url($post->ID);
```
[the remark this answers is not in the document]
Comment (elzadj) [reply core/class-maxi-dynamic-content.php:914]: that's indeed a little confusing, maybe re-assign the $post variable here, and create another one for $post->ID)

**Review by elzadj — CHANGES_REQUESTED** [review]
Almost there, good job!

1. The character limit doesn't work on frontend (checked for the
Biographical Info and Full name, but I guess need to check for everything else too)
2. Current's item data option is acceptable on usual pages, need to hide it there.
3. Open in new tab doesn't make sense for the email links.

**Comment by myroslavve** [comment]
> Almost there, good job!
> 
> 1. The character limit doesn't work on frontend (checked for the
>    Biographical Info and Full name, but I guess need to check for everything else too)

Should be fixed.

> 2. Current's item data option is acceptable on usual pages, need to hide it there.

Added author fields to posts/pages types so we can select current author there, like we discussed in chat. Removed current option from "Author" type (only left when editing author archives). @Olekrut could you check this changes, if this is working properly and UI makes sense.
image

> 3. Open in new tab doesn't make sense for the email links.

Disabled it for author email.

**Review by Olekrut — CHANGES_REQUESTED** [review]
Cool) a few things:

1) guess, need to add a field - static text to author list as well
image
2) type customer > image block > what is should be? ( user's avatar or need to hide? )
image
3) type customer > Country ( billing and shipping )> should we  have ability to see not only short name of country ( guess, yes )
image
4) need to add DC email link for billing email 
image
5) about FSE - get by current archive's item > need to add it to CL for Author archive at least ( guess, because it values are not changing at that moment ) 
image
6) FSE > type Customer > need to add - get by current archive's item
image

**Comment by myroslavve** [comment]
> Cool) a few things:
> 
> 1. guess, need to add a field - static text to author list as well

added

> 2. type customer > image block > what is should be? ( user's avatar or need to hide? )

Hid it for now, will add it if we need it in the future.

> 3. type customer > Country ( billing and shipping )> should we  have ability to see not only short name of country ( guess, yes )

Added fields called full country.

> 4. need to add DC email link for billing email

added

> 5. about FSE - get by current archive's item > need to add it to CL for Author archive at least ( guess, because it values are not changing at that moment )

enabled current for CL, checked, it seems to work right, but it was disabled because of https://github.com/maxi-blocks/maxi-blocks/pull/5528#pullrequestreview-1966580790 first point, so could you check it again pls @Olekrut.

> 6. FSE > type Customer > need to add - get by current archive's item
image

Customers are not authors so shouldn't be there, but I will create an issue to add current option to customers.

**Review by Olekrut — CHANGES_REQUESTED** [review]
1) cant to add link to static text
https://jam.dev/c/cdc032d5-181a-4f45-8e21-bf448e41c3ae
2) need to hide it for customer email as well
image


Everything else is ok for me, Thank you, Great job!

**Review by Olekrut — CHANGES_REQUESTED** [review]
https://jam.dev/c/f7c99435-e666-42a3-9e2a-020604329ede

cant to add link to static text yet

after last fix > link is disappearing for static text

**Comment by myroslavve** [comment]
> https://jam.dev/c/f7c99435-e666-42a3-9e2a-020604329ede
> 
> cant to add link to static text yet
> 
> after last fix > link is disappearing for static text

So DC link doesn't work with custom formats currently, so it was disabled for static text, created a separate issue to allow DC link for static text: https://github.com/maxi-blocks/maxi-blocks/issues/5809

**Review by Olekrut — APPROVED** [review]

**Review by elzadj — CHANGES_REQUESTED** [review]
Looking good, just a final push!

- let's keep the nopluplus rule
- In CL, when Bio info is missing, it gets stuck on the previous content instead of 'No content found' 
Edit-Page-‹-MaxiBlocks-—-WordPress-11-20-2024_02_49_PM
- please add additional checks that rabbit ai asks for
- also pull master, there are some important changes, but they should not affect this PR (hopefully)
- I see a list of site's registered users, not Woo Customers here:
Edit-Page-‹-MaxiBlocks-—-WordPress-11-20-2024_03_23_PM

File: src/components/dynamic-content/index.js [hunk 3]
```diff
@@ -637,15 +648,37 @@ const DynamicContent = props => {
 										onReset={() =>
 											changeProps({
 												'dc-field':
-													fieldOptions[contentType][
+													fieldOptions[contentType][ // TODO: Check if this will work with currentFieldOptions
```
[the remark this answers is not in the document]
Comment (myroslavve) [reply src/components/dynamic-content/index.js]: checked now 😅

**Comment by myroslavve** [comment]
> * In CL, when Bio info is missing, it gets stuck on the previous content instead of 'No content found'
>   Edit-Page-‹-MaxiBlocks-—-WordPress-11-20-2024_02_49_PM

I cannot reproduce this, always returns to "No content found" for me, could you share more details please?

Everything else should be fixed. 👍

**Comment by Olekrut** [comment]
this type is empty - for old page
[dc improvements1.txt](https://github.com/user-attachments/files/17993541/dc.improvements1.txt)
image
image

**Review by Olekrut — CHANGES_REQUESTED** [review]
this type is empty - for old page
[dc improvements1.txt](https://github.com/user-attachments/files/17993541/dc.improvements1.txt)
image


Error
image
image

**Comment by myroslavve** [comment]
> this type is empty - for old page [dc improvements1.txt](https://github.com/user-attachments/files/17993541/dc.improvements1.txt) image
> 
> Error image image

I can't reproduce the error or the "This type is empty" problem, tried on old pages on my local, and created a page on master with the code editor you have provided, both work good for me. Maybe there is something I am missing, if you could record a video in jam would be good.

**Review by Olekrut — CHANGES_REQUESTED** [review]
@elzadj  can re produce it as well with file from previous comment, so
https://jam.dev/c/7e500804-0a01-4088-b936-a9b4bfd5cb0d

**Review by Olekrut — APPROVED** [review]

**Review by Olekrut — CHANGES_REQUESTED** [review]
Change author in CL > click to some block > open CL again > author is backing to first ( automatically )
https://jam.dev/c/8962452b-3864-4343-ae30-c408d7963889

The same with new blocks
https://jam.dev/c/f2a5a7ff-4563-4c6d-b397-772fe407ffda

**Review by Olekrut — APPROVED** [review]

**Review by elzadj — CHANGES_REQUESTED** [review]
The dropdown with woo customers is not working for me:

[Add-New-Page-‹-MaxiBlocks-—-WordPress (1).webm](https://github.com/user-attachments/assets/0a5db868-61ba-4f3b-ac70-c28e4068c379)

**Comment by myroslavve** [comment]
> The dropdown with woo customers is not working for me:
> 
>  Add-New-Page-.-MaxiBlocks-.-WordPress.1.webm

Seems like the problem is happening when there are no customers on the site, at least that's how I could reproduce it.
Fixed for this case, if there are any other cases when it happens, let me know.

**Review by elzadj — CHANGES_REQUESTED** [review]
Getting empty. I have 2 customers with completed orders on the site.

Add-New-Page-‹-MaxiBlocks-—-WordPress-12-10-2024_05_33_PM

**Comment by myroslavve** [comment]
For now, made "current" the only available relation option for customers.

**Review by elzadj — CHANGES_REQUESTED** [review]
Backend and frontend both look weird (my current user is a customer):

[Edit-Page-‹-MaxiBlocks-—-WordPress (3).webm](https://github.com/user-attachments/assets/2ddbf3df-4b8b-4548-8cce-1a14ecffceae)

MaxiBlocks-12-13-2024_11_57_AM

Maybe let's just remove customers?

**Comment by myroslavve** [comment]
> Backend and frontend both look weird (my current user is a customer):
> 
>  Edit-Page-.-MaxiBlocks-.-WordPress.3.webm 
> MaxiBlocks-12-13-2024_11_57_AM
> 
> Maybe let's just remove customers?

Yes, removed them for now, it would be hard to fix your issue, so better do it if we will be implementing customers fully in another issue.

**Review by elzadj — APPROVED** [review]

**Comment by elzadj** [comment]
Looks good
@Olekrut check pls, and then we merge

**Review by Olekrut — APPROVED** [review]

**Review by elzadj — APPROVED** [review]

**Pull request merged**